src/osp/id_registry.h

```diff
@@ -0,0 +1,349 @@
+                    "Reached max capacity with automatic resizing disabled");
+        }
+        else
+        {
+            reserve(std::max(capacity() + count, capacity() * 2));
+        }
+    }
+
+    m_deleted.take<IT_T, ID_T>(out, count);
+}
+
+//-----------------------------------------------------------------------------
+
+/**
+ * @brief Long term reference counted storage for IDs
+ */
+template<typename ID_T, typename REG_T>
+class IdStorage
+{
+    friend REG_T;
+
+public:
+    IdStorage() : m_id{ id_null<ID_T>() } { }
+    IdStorage(IdStorage&& move) = default;
+    ~IdStorage() { assert( ! has_value() ); }
+
+    // Delete copy
+    IdStorage(IdStorage const& copy) = delete;
+    IdStorage& operator=(IdStorage const& copy) = delete;
+
+    // Allow move assignment only if there's no ID stored
+    IdStorage& operator=(IdStorage&& move)
+    {
+        assert( ! has_value() );
+        m_id = std::exchange(move.m_id, id_null<ID_T>());
+        return *this;
+    }
+
+    operator ID_T() const noexcept { return m_id; }
+    constexpr ID_T value() const noexcept { return m_id; }
+    constexpr bool has_value() const noexcept { return m_id != id_null<ID_T>(); }
+
+private:
+    IdStorage(ID_T id) : m_id{ id } { }
+
+    ID_T m_id;
+
+}; // class IdStorage
+
+//-----------------------------------------------------------------------------
+
+/**
+ * @brief An IdRegistry that uses IdStorages to uniquely store Ids
+ */
+template<typename ID_T, bool NO_AUTO_RESIZE = false>
+class UniqueIdRegistry : IdRegistry<ID_T, NO_AUTO_RESIZE>
+{
+    using base_t = IdRegistry<ID_T, NO_AUTO_RESIZE>;
+public:
+
+    using Storage_t = IdStorage<ID_T, UniqueIdRegistry<ID_T, NO_AUTO_RESIZE> >;
+
+    using IdRegistry<ID_T, NO_AUTO_RESIZE>::IdRegistry;
+    using base_t::capacity;
+    using base_t::size;
+    using base_t::reserve;
+    using base_t::exists;
+
+    [[nodiscard]] Storage_t create()
+    {
+        return base_t::create();
+    }
+
+    void remove(Storage_t& rStorage)
+    {
+        base_t::remove(rStorage);
+        rStorage.release();
+    }
+}; // class UniqueIdRegistry
+
+
+//-----------------------------------------------------------------------------
+
+template<typename COUNT_T = uint8_t>
+class RefCount : std::vector<COUNT_T>
+{
+    using base_t = std::vector<COUNT_T>;
+public:
+
+    RefCount() = default;
+    RefCount(RefCount&& move) = default;
+    RefCount(size_t capacity) { resize(capacity); };
+
+    // Delete copy
+    RefCount(RefCount const& copy) = delete;
+    RefCount& operator=(RefCount const& copy) = delete;
+
+    // Allow move assign only if all counts are zero
+    RefCount& operator=(RefCount&& move)
+    {
+        assert(isRemainingZero(0));
+        base_t(std::move(move));
+        return *this;
+    }
+
+    ~RefCount()
+    {
+        // Make sure ref counts are all zero on destruction
+        assert(isRemainingZero(0));
+    }
+
+    bool isRemainingZero(size_t start) const noexcept
```

Comment:
Intentional formatting change? Is this something to do with the fact that it's a predicate?

Reply:
I have been cursed by the demons of javascript and have accidentally camel cased a function.

Reply:
It happens to even the most dogmatic of us